theano/scan_module/scan_opt.py

```diff
@@ -294,6 +326,14 @@ def add_to_replace(y):
                 # We have a candidate node to removable
                 # Step 1. Reconstruct it on outside
                 to_remove_set.add(nd)
+                node_clients = retrieve_clients_fromnodes(nd)
+
+                # Add the nodes only if they are not already
+                # in the bookkeeper's list
+                nd_cli_cleaned = filter(lambda x: x not in local_fgraph_bookkeeper,
+                                        node_clients)
+                local_fgraph_bookkeeper.extend(nd_cli_cleaned)
```

Comment:
I think we have an ordered set in Theano
Le 10 sept. 2015 19:21, "Caglar" <notifications@github.com> a écrit :

> In theano/scan_module/scan_opt.py
> <https://github.com/Theano/Theano/pull/3375#discussion_r39226619>:
>
> > @@ -294,6 +326,14 @@ def add_to_replace(y):
> >                  # We have a candidate node to removable
> >                  # Step 1. Reconstruct it on outside
> >                  to_remove_set.add(nd)
> > +                node_clients = retrieve_clients_fromnodes(nd)
> > +
> > +                # Add the nodes only if they are not already
> > +                # in the bookkeeper's list
> > +                nd_cli_cleaned = filter(lambda x: x not in local_fgraph_bookkeeper,
> > +                                        node_clients)
> > +                local_fgraph_bookkeeper.extend(nd_cli_cleaned)
>
> Yeah, but in that case I would need to use an OrderedSet(which doesn't
> exist in python itself, but there are other ways to implement it), which is
> not that efficient either. I think, the size of the bookkeeper shouldn't be
> very large, thus N in O(N) is not very large probably less than the
> constant in O(1). But probably worth trying out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3375/files#r39226619>.
>